gnome/gobject-introspection/files/patch-fix-scanner-in-build-execution.diff

```diff
@@ -0,0 +1,61 @@
+--- giscanner/ccompiler.py.orig
++++ giscanner/ccompiler.py
+@@ -184,7 +184,10 @@
+         if os.name == 'nt':
+             runtime_path_envvar = ['LIB', 'PATH']
+         else:
+-            runtime_path_envvar = ['LD_LIBRARY_PATH', 'DYLD_FALLBACK_LIBRARY_PATH']
++            if sys.platform == "darwin":
++                runtime_path_envvar = ['DYLD_LIBRARY_PATH']
++            else:
++                runtime_path_envvar = ['LD_LIBRARY_PATH']
+             # Search the current directory first
+             # (This flag is not supported nor needed for Visual C++)
+             args.append('-L.')
+--- giscanner/dumper.py.orig
++++ giscanner/dumper.py
+@@ -236,8 +236,25 @@
+ 
+         args.extend(sources)
+ 
+-        pkg_config_libs = pkgconfig.libs(self._packages,
++        pkg_config_libs, pkg_config_libs_only_L = pkgconfig.libs(self._packages,
+                                          msvc_syntax=self._compiler.check_is_msvc())
++        this_L = []
++        for _pc_libs_L in pkg_config_libs_only_L:
++            this_L.append(_pc_libs_L[len('-L'):])
++
++        if os.name == 'nt':
++            runtime_path_envvar = ['LIB', 'PATH']
++        else:
++            if sys.platform == "darwin":
++                runtime_path_envvar = ['LIBRARY_PATH','DYLD_FALLBACK_LIBRARY_PATH']
```

Comment:
I don't have experience with non-Mac operating systems, but on macOS, at runtime, the dynamic library loader will look for libraries first in the locations specified in `DYLD_LIBRARY_PATH`, then in the locations baked into the executable or library, then in locations specified in `DYLD_FALLBACK_LIBRARY_PATH`. I agree with Ken that we don't have any need to look in `DYLD_FALLBACK_LIBRARY_PATH` at any time.

`LIBRARY_PATH` is not used at runtime. Specifying a path in `LIBRARY_PATH` is similar to specifying it with an `-L` flag: it is used by the linker at link time to find libraries. I'm not sure what dumper.py does. Does it invoke the linker? If so, then maybe it does need to set environment variables that will be used by the linker, and it's just a case of a misnamed variable (`runtime_path_envvar`).

Reply:
This specific chunk of `dumper.py` is setting up the link command and environment. Hence using `LIBRARY_PATH` is the way to go for Darwin / macOS and why I split this setting off from other non-Windows settings. Any `DYLD` settings here will not impact the link.

That said, settings here are carried forth throughout the rest of the program. Hence minimizing the settings here is best. I will remove the `FALLBACK` portion, as I've tested this both with and without it and it works fine both ways, and hence that portion is not required.

Reply:
FYI `poppler`s CMake script adds on the paths for the local build and source to find both headers and libraries. gO-I correctly sets those paths for the build and link commands. What I'm fixing here is the use of "system" library search path, in this case defined as paths that are not in the build paths somewhere.

The overall code flow here is to glean library linkage information using `pkg-config` and add it to the general link command. This is fine -except- for the inclusion of any `-L` "system" paths, which are better off being in `LIBRARY_PATH` instead of on the link command. This be because prior `-l` libraries are in-build -- with the proper `-L` to find them -- and those libraries might have an installed version already in the "system" paths. Hence moving the "system" library paths from `-L` to `LIBRARY_PATH` allows the linker find the in-build version instead of the already-installed "system" version ... which is what this `poppler` build issue is all about in the first place!

Reply:
Reading dumper.py now, I see that this code you're changing is inside a function called `_link` which does call the linker.

Before it was asking pkg-config for all flags, including `-L` flags, and adding those to project-specific flags, including `-L` flags, and then calling the linker with all those args. Now you've changed it so that the pkg-config `-L` flags get changed into paths in `LIBRARY_PATH` instead, which should work the same except for the very old versions of clang that don't support `LIBRARY_PATH` that we don't care about anymore.

What I don't see yet is where those args are carried forward to the rest of the program. Where's that?

Reply:
> What I don't see yet is where those args are carried forward to the rest of the program. Where's that?

Not sure what you mean here. I'll take a guess:

Once the `os.environ` is set, those variables are carried forth inside Python until they are either changed / unset or Python exits. Unlike local variables inside Python (or C/C++ or most languages), these variables are set for the overall runtime environment and hence if we don't want them carry forth we'll need to save the old environment, modify it, and restore it at the end. Not difficult to do, but I also don't see a great need for doing so.

Reply:
>> What I don't see yet is where those args are carried forward to the rest of the program
>
> Not sure what you mean here

Well you said: 

> settings here are carried forth throughout the rest of the program

But I should probably stop asking questions and just test your changes first.

Reply:
> > settings here are carried forth throughout the rest of the program

Yes ... as noted in my reply and augmenting here: Unlike local variables inside Python (or C/C++ or most languages), these variables are set for the overall Python runtime environment. Hence, setting them here will carry them forth beyond just this specific function into the code that called it, and the code that called that code, etc -- until either the runtime environment is changed or Python returns from execution.

One option we have is to modify the code to save the runtime environment, change it, do the link, then restore the runtime environment back to the state it was in at the start of the function. Doing this won't be too difficult, but I also don't see doing it as necessary.

> But I should probably stop asking questions and just test your changes first.

Testing the changes is definitely in order here. The code isn't the most obvious. I added a ton of debug `print` statements to see what was going on throughout the code execution, to get to the point of these changes. I'm confident that these changes do the trick for MacPorts' needs. But any testing I did cannot be exhaustive & hence I might have missed something. Thus, yes, I appreciate others testing / verifying that these changes work for both `poppler` as well as don't impact other ports that use gO-I .